src/main/java/org/apache/mesos/scheduler/DefaultTaskKiller.java

```diff
@@ -0,0 +1,95 @@
+package org.apache.mesos.scheduler;
+
+import org.apache.mesos.Protos;
+import org.apache.mesos.SchedulerDriver;
+import org.apache.mesos.scheduler.recovery.TaskFailureListener;
+import org.apache.mesos.state.StateStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * This class is a default implementation of the TaskKiller interface.
+ */
+public class DefaultTaskKiller implements TaskKiller {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final StateStore stateStore;
+    private final TaskFailureListener taskFailureListener;
+
+    private final Object restartLock = new Object();
+    private List<Protos.TaskInfo> tasksToRestart = new ArrayList<>();
+    private final Object rescheduleLock = new Object();
+    private List<Protos.TaskInfo> tasksToReschedule = new ArrayList<>();
+
+    public DefaultTaskKiller(StateStore stateStore, TaskFailureListener taskFailureListener) {
+        this.stateStore = stateStore;
+        this.taskFailureListener = taskFailureListener;
+    }
+
+    @Override
+    public void killTask(String taskName, boolean destructive) {
+        Optional<Protos.TaskInfo> taskInfoOptional = stateStore.fetchTask(taskName);
+        if (!taskInfoOptional.isPresent()) {
+            logger.warn("Attempted to kill unknown task: " + taskName);
+            return;
+        }
+
+        Protos.TaskInfo taskInfo = taskInfoOptional.get();
+        if (!stateStore.fetchStatus(taskName).getState().equals(Protos.TaskState.TASK_RUNNING)) {
+            logger.warn("Task is not running, no need to kill: " + taskInfo);
+            return;
+        }
+
+        if (destructive) {
+            synchronized (rescheduleLock) {
+                logger.info("Scheduling task to be killed destructively: " + taskInfo);
+                tasksToReschedule.add(taskInfo);
```

Comment:
Throw IllegalArgumentException (or just `warn`) here if taskInfo is null. Then the null checks in processTasksTo[Restart|Reschedule] below can be removed.